ddcci-multi-monitor@tim-we/files/ddcci-multi-monitor@tim-we/applet.js

```diff
@@ -188,6 +195,46 @@ class DDCMultiMonitor extends Applet.IconApplet {
             this.updateMenu();
         }
     }
+
+    // Change the brightness when scrolling on the icon
+    _onScrollEvent(actor, event) {
+        let direction = event.get_scroll_direction();
+        if (direction == Clutter.ScrollDirection.SMOOTH) { // Prevents being triggered twice
+            return
+        }
+
+        if (direction == Clutter.ScrollDirection.DOWN) {
+            this.monitors.forEach((monitor) => {
+                monitor.brightness = Math.max(0, monitor.brightness - BRIGHTNESS_ADJUSTMENT_STEP); // Decrease brightness by the step amount, making sure to not go below 0
+                // Tried making this into a function, it did not go well
+                this.set_applet_tooltip("Brightness: " + monitor.brightness + "%"); 
+                this._applet_tooltip.show();
+                setTimeout(() => { // Wait before actually changing the brightness, to prevent weirdness when scrolling multiple steps
+                    monitor.setBrightness(monitor.brightness); 
+                    setTimeout(() => {
+                        this._applet_tooltip.hide();
+                        this.set_applet_tooltip(DEFAULT_TOOLTIP);
+                    }, 1000);
+                }, 180);
```

Comment:
> to prevent weirdness when scrolling multiple steps

Could you elaborate on that? I know that ddc/ci behaves weird when you issue a new command before a previous one has completed. Maybe we should update `setBrightness` accordingly. It could store the promise in a private field of `Monitor` and append the new operation with `.then(...)`.
I'm only assuming thats the issue here and I think that would be a cleaner solution.

Reply:
> Could you elaborate on that?

Without the pause, the screen brightness would jump around, often ending on the wrong brightness. I assume this is because of ddc/ci sending the brightness changes in a different order?

> Maybe we should update `setBrightness` accordingly. It could store the promise in a private field of `Monitor` and append the new operation with `.then(...)`.

While I did have to do some searching to figure out what that means, that's a much better idea than what I did!

My new code for the brightness setting:
```js
class Monitor {
    #promises;
    constructor(index, name, bus) {
        this.index = index;
        this.name = name;
        this.brightness = 50;
        this.bus = bus;
        this.menuLabel = null;
        this.menuSlider = null;
        this.#promises = Promise.resolve();
    }
```
...
```js
    setBrightness(value) {
        this.brightness = Math.round(value);
        this.updateMenu();
        this.#promises = this.#promises.then(() => {
            return new Promise((resolve, reject) => {
                Util.spawnCommandLineAsync(
                    `ddcutil --bus=${this.bus} setvcp 10 ${this.brightness}`,
                    resolve,
                    reject
                );
            });
        });
    }
```
...
```js
class DDCMultiMonitor extends Applet.IconApplet {
```
...
```js
        if (direction == Clutter.ScrollDirection.DOWN) {
            let tooltipMessage = this.monitors.map(monitor => {
                monitor.brightness = Math.max(0, monitor.brightness - BRIGHTNESS_ADJUSTMENT_STEP);
                monitor.setBrightness(monitor.brightness);
                return `${monitor.name}: ${monitor.brightness}%`;
            }).join("\n");

            this.set_applet_tooltip(tooltipMessage)
            this._applet_tooltip.show()
            setTimeout(() => {
                this._applet_tooltip.hide();
                this.set_applet_tooltip(DEFAULT_TOOLTIP);
            }, 2500);
            
        }
        else if (direction == Clutter.ScrollDirection.UP) {
```


Should I push this into the pull request?

Reply:
Yes, looks good to me. Only thing that we should change is the `setTimeout` thing for hiding the tooltip. If I understand correctly this will get scheduled for every scroll step. So if I scroll up and down for 2.5s weird things will start to happen. `setTimeout(...)` returns an id for that timeout which can be used to cancel the timeout. See https://developer.mozilla.org/en-US/docs/Web/API/clearTimeout#examples
I would suggest to create a method for scheduling the hiding of the tooltip and this method could handle the clearing of the timeout as well as creating a new timeout. 

> While I did have to do some searching to figure out what that means

Sorry about that :D Anyway, I hope you learned something, I think promises are really cool and a handy tool for async stuff.

Thanks for your contribution to this applet!

Reply:
Also feel free to just commit & push these changes to this branch, instead of a long comment like this. That makes it easier to review.